### ISSUE TITLE
NLDI related use case: priority date order

### ISSUE THREAD

**Issue by amabdallah** [WSWCWaterDataExchange/WestDAAT#22]
For Frank the Farmer, Tony once mentioned this point and I love it. I'm not sure how hard it would be to code. Suppose the NLDI API returns 200 WaDE POD sites. Each site has its priority date from its parent water right. Our WestDAAT code sorts the sites based on priority date from the oldest to the newest. Then Assign he oldest one a value of 1 and the newest one a value of 200. If two priority dates are equal to the same year and month, then assign them the same order e.g., 5. WestDAAT should show the numeric value as a label. The "label" should be included as part of an export function (i.e., include it in the downloaded data)
 
The order from 1-99 is a normalized index across potentially many places. For now, it would be easier to have the highest order depending on being whatever the highest number of sites we have. 

![image](https://user-images.githubusercontent.com/3268971/141190826-693916e4-0cd8-4742-b4b9-3a0724c5eb75.png)

![image](https://user-images.githubusercontent.com/3268971/141191335-d65d070a-eed2-4645-a567-577c158e893d.png)

**Comment by rwjam** [comment]
Neat idea.  This sounds like an "visualize Priority Date" ability request.  Might be a good idea to add this as a new option for the legend colors, as I can see this use case being used in other instances besides NLDI.

![image](https://user-images.githubusercontent.com/17555073/141510831-7f0e9e4d-ec4c-41c0-84d2-beee420f81b4.png)

Note: 
- This is definitely on DPL to provide us with the answer of how hard it will be to code / how much $$ this will cost to develop.  As the priority date order is highly dependent on the search results made by the user, I can't see of a way to store this information ahead of time on our end.

**Comment by amabdallah** [comment]
I agree that DPL will help us estimate this. We need to share a dummy or real data sample with them showing this priority date order. One column as 50 priority date values sorted from old to new and the second column gives them incremental integer values starting from 1. 
I like the idea of adding priority date to the legend options and since there are virtually hundreds of unique priority date years, we need to make the legend in decade range e.g.., (1880-1889, 1890-1899) which would cut the legend items down to probably <~15. That being said, what would be the use case or insight out of this filter? We might need to pre-assign this decade value to each priority date and store it in the database. I see that in can simplify the growth animation. In the end, this is probably would be on the nice to have feature compared to the other legend filters.

**Comment by rwjam** [comment]
> We need to share a dummy or real data sample with them showing this priority date order.

- We can try this.
- These numbers would be made up though, as I can't see how we would store those numbers ahead of time to cut down development cost.  They have to be built on the fly / unique with each filter request, which is a WestDAAT custom feature.

> ...virtually hundreds of unique priority date years, we need to make the legend in decade range e.g.., (1880-1889, 1890-1899) which would cut the legend items down to probably <~15...

- Agreed.  That would simplify the visual.
- I was thinking something like this (see image), where you would provide some sort of 'rank' with like a youngest-oldest legend, then provide like a number tag for each site.
![image](https://user-images.githubusercontent.com/17555073/141516428-39d9ba8f-352f-4810-9a4d-6de597ec7ff3.png)


> That being said, what would be the use case or insight out of this filter? 

- Gary the Governor / Stan the State Engineer might want to see priority hierarchy across several river basins in their state  As I understand, the NLDI tool only reports the one river basin of interest back.

> ...pre-assign this decade value to each priority date and store it in the database....

- I wouldn't think you would need to pre-assign a value ahead of time.  Priority Date is stored as a timeseries data type in a YYYY-MM-DD format, which be sorted, then you can assign a rank to that sort, then show the rank as a visual.  You could do the sort / rank by year to simplify it.
You could even add a hierarchy to the return rank, so results would might look like, 5 (oldest), 4a, 4b, 4c, 3, 2a, 2b, 1.

**Comment by amabdallah** [comment]
Good idea. The rank can give an insight at a higher level. From a farmer perspective, they want to know their exact order (e.g., how many water rights upstream are ahead or behind of my priority date?" So this is why 1- 100 order is helpful in that case. 

This is a strictly an NLDI post-processed result that is dynamic to any selected point on the stream network. So the order may change based on how many water rights are upstream of a selected point. We could run an NLDI call for each supported river basin at the mouth of the river (the most downstream point) and then store the absolute priority date order in the database. This would be a lot to ask the NLDI call to return all the sites in a basin and I'm not sure if NLDI can handle that. 

For now, let's support the farmer perspective use case where the priority order is handled/generated on the fly after the NLDI results come back. Showing a label with the number should be fine for now. We do need to export that generated label if the user wants to download the NLDI results. I hope this helps and let's chat more over Zoom
